test/iterators/bfs.jl

```diff
@@ -35,7 +35,7 @@
         end
     end
     nodes_visited = collect(BFSIterator(g2, [1, 6]))
-    @test nodes_visited == [1, 2, 3, 6, 5, 7, 4]
+    @test nodes_visited == [1, 6, 2, 3, 5, 7, 4]
```

Comment:
If these tests are sensitive to the ordering at each level, which is an implementation detail, can you rework them to make them independent?

Reply:
I'm a bit unsure if the ordering on each level could be considered an implementation detail, if this is it we could speed-up the running time by 2x by ordering the level nodes (for cache locality reasons I presume), but this is actually wrong because you want to follow strictly how bfs works, which is to look for each of the neighbors of a certain node in the previous level and just then go to the next node of the previous level

Reply:
If you consider two graph structures where the neighbors of each vertex make up the same (mathematical) set but are stored in different orders, the BFS algorithm will return different things for `node_visited`. And they will both be correct. So our tests should be agnostic to that

Reply:
yes I mean I think you are right that we should be agnostic to that, just to clarify what I mean:
          
             0 
          /      \
        1         2
       /  \     /  \
      3    4    5    6

given a graph like this if we start at node 0 it is okay to have e.g. `0, 1, 2, 3, 4, 5, 6` or `0, 2, 1, 5, 6, 3, 4` but not `0, 1, 2, 5, 6, 3, 4` or `0, 1, 2, 5, 3, 4, 6`.

But I think this what you are actually saying in your last comment, so we need to have tests which are okay with all acceptable versions.

Reply:
Exactly. It's a bit of a pain so I'm not making it strictly necessary for the PR to be merged, but essentially in your example we would want to check that the returned vector has the form `[.|..|....]` where the first subset is `{0}`, the second is `{1, 2}` and the third is `{3, 4, 5, 6}` but in any order

Reply:
that would be actually not okay though I think, because for example `0, 1, 2, 5, 6, 3, 4` is not acceptable

Reply:
Your definition iterates sligthly differently than what I had in mind...but it's totally okay I wanted just to understand which one was preferable and indeed parallelizing the algorithm effectively would require to drop mine. So let's go with yours, I think that we can also get a 2x speed-up by sorting with that :-)

Reply:
I'm slightly biased by a recent bachelor project I supervised on... parallel BFS ;) check out the repo of my interns https://github.com/KassFlute/ParallelGraphs.jl for a multithreaded and even BLAS-ified version of BFS that is much faster than the one here! ping @KassFlute and @AntoineBut

Reply:
Gentle ping @Tortar if you want to adjust the tests so that I can merge while it's still fresh in our minds

Reply:
should be okay now :+1:

Reply:
thanks!